### PR TITLE
Move service levels information to `system_distributed_everywhere` keyspace

### LINKED ISSUE #12706: system_distributed.service_levels configuration table should be moved to system_distributed_everywhere as it is too fragile
The `service_levels` table which is part of the `system_distributed` keyspace was proven to be too fragile. This table is polled by every node every 10 seconds with a scan query (which doesn't have speculative retry) with `CL=ONE`.
This combination causes two problems:
1. Availability - since `system_distributed` has a constant replication factor (3 in the time of writing) , the larger the cluster, more and more nodes are involved in the scan query. Since we don't have speculative retry for scan queries (by design) if one of the involved nodes fails to reply, it fails the whole scan query causing the time of a new configuration propagation to be very unpredictable for an overloaded cluster.
2. Non linearity - when the table has inconsistencies (for the normal reasons which Scylla is designed to deal with), it can happen that every poll will result in a different observed configuration. Until the table is repaired, there will be flactuations where a specific added service level can appear in some of the polls and not appear in others or it will appear with different configuration, this in turn will cause unnecesairy and incorrect configuration updates on the node, making it alternate between those observed configrations back and forth. This, besides from the fact that it loads the system, creates unpredictability.

For those two reasons we need to:
1. Move the `service_levels` table to the `system_distributed_everywhere` keyspace.
2. Consider queriying the table sometimes (with some probability or evry x polls) with 'CL=TWO' (or CL='ALL') in order to have anti entropy and reduce the need for repairing this table on a regular basis.

Chalenges:
There are 2 main challenges for this change one of which is common to all users of `EverywhereStrategy`:
1. Common - Repair of a table distributed everywhere is fragile, and the probablility for it to fail correlates tightly with the cluster size (nodes and shards number).
2. Secific - All upgrade paths should be crefully considered and tested since the user might loose some of the configuration if the transition is not done correctly.

### PR DESCRIPTION
**Pull request by Jadw1**
This PR adds upgrade mechanism to change configuration table from `system_distributed.service_levels` to `system_distributed_everywhere.service_levels_v2`.

The Scylla-code part is done, but this is draft PR because I'm writing test to check if upgrade works correctly during rolling upgrade (dtests).

The upgrade is protected under `SERVICE_LEVELS_TABLE_V2` feature.

---

Until the upgrade can be performed, `service_level_controller`'s distributed data accessor works as follow:
- reads are done to old table (`system_distributed.service_levels`)
- writes are done to both of the tables

The upgrade process copies the data by copying local mutations of the old table and applying them locally to the new one.
After the mutations are copied, data accessor will use only V2 table.

All of the queries (both before and after upgrade) are done with consistency level `LOCAL_ONE`.

---

Fixes #12706

### CODE REVIEW

File: service/qos/service_level_controller.cc [hunk 1]
```diff
@@ -312,6 +329,9 @@ void service_level_controller::update_from_distributed_data(std::chrono::duratio
             }
         });
     }
+    if (_global_controller_db->distributed_controller_upgrade.available()) {
```
[the remark this answers is not in the document]
Comment (Jadw1) [reply service/qos/service_level_controller.cc:332]: It is checked inside the loop. If the upgrade already happened, then I'll only wait for `_sl_data_accessor->is_data_source_ready()` and exit upgrade loop

File: service/qos/service_level_controller.cc [hunk 2]
```diff
@@ -22,29 +24,29 @@ sstring service_level_controller::default_service_level_name = "default";
...
     if (this_shard_id() == global_controller) {
         _global_controller_db = std::make_unique<global_controller_data>();
         _global_controller_db->default_service_level_config = default_service_level_config;
     }
 }
 
-future<> service_level_controller::add_service_level(sstring name, service_level_options slo, bool is_static) {
-    return container().invoke_on(global_controller, [=] (service_level_controller &sl_controller) {
+future<> service_level_controller::add_service_level(sstring name, service_level_options slo, bool is_static) {  
+    co_return co_await container().invoke_on(global_controller, [=] (service_level_controller &sl_controller) {
```
[the remark this answers is not in the document]
Comment (Jadw1) [reply service/qos/service_level_controller.cc]: reverted

File: service/qos/service_level_controller.cc [hunk 3]
```diff
@@ -22,29 +24,29 @@ sstring service_level_controller::default_service_level_name = "default";
...
         return with_semaphore(sl_controller._global_controller_db->notifications_serializer, 1, [&sl_controller, name, slo, is_static] () {
            return sl_controller.do_add_service_level(name, slo, is_static);
         });
     });
 }
 
 future<>  service_level_controller::remove_service_level(sstring name, bool remove_static) {
-    return container().invoke_on(global_controller, [=] (service_level_controller &sl_controller) {
+    co_return co_await container().invoke_on(global_controller, [=] (service_level_controller &sl_controller) {
```
[the remark this answers is not in the document]
Comment (Jadw1) [reply service/qos/service_level_controller.cc]: reverted

File: db/system_distributed_keyspace.cc [hunk 4]
```diff
@@ -834,9 +900,9 @@ future<qos::service_levels_info> system_distributed_keyspace::get_service_levels
     });
 }
```
[the remark this answers is not in the document]
Comment (Jadw1) [reply db/system_distributed_keyspace.cc:912]: fixed, queries to old table are with `ONE`, to the new table with `LOCAL_ONE`

File: db/system_distributed_keyspace.cc [hunk 5]
```diff
@@ -152,6 +155,19 @@ schema_ptr service_levels() {
     return schema;
 }
```
[the remark this answers is not in the document]
Comment (Jadw1) [reply db/system_distributed_keyspace.cc:157]: done

**Comment by Jadw1** [comment]
Changelog:
- performing upgrade from data accessor level
- V2 table is created by copying columns from the old schema
- different CL for old and new tables

**Comment by Jadw1** [comment]
@eliransin ping for review

**Comment by mykaul** [comment]
> @eliransin ping for review

@eliransin - is it going to get into 5.3?

File: db/system_distributed_keyspace.hh [hunk 6]
```diff
@@ -44,6 +45,7 @@ public:
 
     static constexpr auto VIEW_BUILD_STATUS = "view_build_status";
     static constexpr auto SERVICE_LEVELS = "service_levels";
+    static constexpr auto SERVICE_LEVELS_V2 = "service_levels_v2";
```
Comment (eliransin) [inline db/system_distributed_keyspace.hh:48]: I don't have any particular problem with this name, but just for completeness sake, why did you choose to go with a different table name?

File: service/qos/service_level_controller.cc [hunk 7]
```diff
@@ -88,15 +90,23 @@ future<> service_level_controller::drain() {
...
-        } catch (const sleep_aborted& ignored) {
-        } catch (const exceptions::unavailable_exception& ignored) {
-        } catch (const exceptions::read_timeout_exception& ignored) {
-        }
+
+    future<> data_accessor_stop_fut = make_ready_future<>();
+    if (_sl_data_accessor) {
+        data_accessor_stop_fut = _sl_data_accessor->stop();
+    }
+    return data_accessor_stop_fut.then([this] {
```
Comment (eliransin) [inline service/qos/service_level_controller.cc:98]: I think that the order between stopping the dta_accessor and the distributed update should be switched. The reason is that once the distributed update loop is stopped, the are going to be very little (at current code state none) calls to the data_accessor as the distributed_update is one of it's only users.

File: service/qos/service_level_controller.hh [hunk 8]
```diff
@@ -49,10 +49,18 @@ class service_level_controller : public peering_sharded_service<service_level_co
...
     public:
+        virtual future<> init() = 0;
+        virtual future<> stop() = 0;
+
+        // virtual bool is_data_source_ready() const = 0;
         virtual future<qos::service_levels_info> get_service_levels() const = 0;
         virtual future<qos::service_levels_info> get_service_level(sstring service_level_name) const = 0;
         virtual future<> set_service_level(sstring service_level_name, qos::service_level_options slo) const = 0;
         virtual future<> drop_service_level(sstring service_level_name) const = 0;
+
```
Comment (eliransin) [inline service/qos/service_level_controller.hh:60]: Did those commented out functions left here intentionally?
(I guess not and they should probably be cleaned up :slightly_smiling_face: )

File: service/qos/standard_service_level_distributed_data_accessor.cc [hunk 9]
```diff
@@ -8,27 +8,92 @@
...
 
 future<> standard_service_level_distributed_data_accessor::drop_service_level(sstring service_level_name) const {
-    return _sys_dist_ks.drop_service_level(service_level_name);
+    if (!_v2_upgraded) {
+        co_await _sys_dist_ks.drop_service_level(service_level_name);
+    }
+    co_await _sys_dist_ks.drop_service_level_v2(service_level_name);
+}
+
+future<> standard_service_level_distributed_data_accessor::upgrade_loop() {
```
Comment (eliransin) [inline service/qos/standard_service_level_distributed_data_accessor.cc:77]: This is a little bit sub optimal but we can do it as a followup, I would expect this upgrade loop to only start once the feature is enabled. (registering a notification on it).
But right now the current mechanism cripples us anyway (this started() function), I would
expect it to be a future or at least to have a future version so we can just wait on it, of course that then we would need to handle aborts differently.
So it is definitely good enough given the other code we interact with, just something to keep in mind :slightly_smiling_face:

**Review by eliransin — APPROVED** [review]
LGTM to me except for some minor comments.
Please have another look in case my comments make sense to you.

**Comment by avikivity** [comment]
Could we do it differently?
1. change the replication factor to exactly match the topology
2. run repair
3. change the replication strategy

there's a hole between 1 and 2, since at that point reads can return incorrect data.

**Comment by avikivity** [comment]
I believe you need three cluster features:
 - one to get all nodes to agree that an upgrade is needed and can start
 - another to get all nodes to agree to switch to reading from the new table
 - yet another to get all nodes to stop writing to the old table

Shouldn't we make this raft-managed? @kbr-scylla

**Comment by kbr-scylla** [comment]
> Shouldn't we make this raft-managed? @kbr-scylla

If you mean making `service_levels` table managed by group 0,  then yes, we probably should.
cc @gleb-cloudius 

There was a similar story with CDC internal tables: they were in `system_distributed` initially, then they were in `system_distributed_everywhere` (that's when the keyspace was created), now we're moving them to `system` and it's managed by group 0.

**Comment by gleb-cloudius** [comment]
> > Shouldn't we make this raft-managed? @kbr-scylla
> 
> If you mean making `service_levels` table managed by group 0, then yes, we probably should. cc @gleb-cloudius
> 

May favorite question to answer when we consider if something should be a part of group0: what problem will it solve.

Service level config should not be serialized with topology for correctness right? What about DDL? 

> There was a similar story with CDC internal tables: they were in `system_distributed` initially, then they were in `system_distributed_everywhere` (that's when the keyspace was created), now we're moving them to `system` and it's managed by group 0.

CDC is serialized with the topology.

**Comment by Jadw1** [comment]
> Could we do it differently?
> 
> 1. change the replication factor to exactly match the topology
> 2. run repair
> 3. change the replication strategy
> 
> there's a hole between 1 and 2, since at that point reads can return incorrect data.

I think we can avoid the hole by not messing with RF of the old table and repairing the new one.

So the upgrade on each node would look like:
1. Copy local mutations to new table
2. Perform repair of the new table (so the last node will sync the new table to finally look exactly like the old one)
3. Announce feature bit saying "I've copied my data to the new table"
(here node still writes to both tables and read from the old)
4. After the cluster feature is on, start reading from the new table and announce feature bit saying "I'm reading from the new table"
5. After this feature is on, stop writing to the old table

File: service/qos/service_level_controller.hh (hunk 8 above)
Comment (Jadw1) [reply service/qos/service_level_controller.hh:60]: yep, it's a leftover I haven't noticed

**Comment by eliransin** [comment]
@avikivity 
We can probably avoid also the repair by using "mutate_internal" instead of "mutate_locally" but I need to verify this.
But as @Jadw1 wrote we will need to take the phased approach you suggested to avoid a window where the configuration might be out of sync. It is pretty straight forward though.

> Shouldn't we make this raft-managed? @kbr-scylla

I think that eventually it might be beneficial to move all cluster wide configurations to be managed under group0, but this will require generalization for this process, one more similarity that I can mention with that regard is the system_auth keyspace which we would like to move to `EverywhereReplication`. I personally prefer the two phase approach since it is unknown when and if we will generalize. I am convinced however that after the move to EverywhereReplication it will be a smoother transition to group0 since the behavior will be very similar.

**Comment by kbr-scylla** [comment]
> one more similarity that I can mention with that regard is the system_auth keyspace which we would like to move to EverywhereReplication

cc @kostja

**Comment by Jadw1** [comment]
> > Could we do it differently?
> > 
> > 1. change the replication factor to exactly match the topology
> > 2. run repair
> > 3. change the replication strategy
> > 
> > there's a hole between 1 and 2, since at that point reads can return incorrect data.
> 
> I think we can avoid the hole by not messing with RF of the old table and repairing the new one instead.
> 
> So the upgrade on each node would look like:
> 
> 1. Copy local mutations to new table
> 2. Perform repair of the new table (so the last node will sync the new table to finally look exactly like the old one)
> 3. Announce feature bit saying "I've copied my data to the new table"
>    (here node still writes to both tables and reads from the old)
> 4. After the cluster feature is on, start reading from the new table and announce feature bit saying "I'm reading from the new table"
> 5. After this feature is on, stop writing to the old table
> 
> Steps 1 and 2 can be done without any cluster feature since in only touches local data and there is no need to wait for other nodes.

@avikivity any comments/thoughts?

**Comment by avikivity** [comment]
> > Could we do it differently?
> > 
> > 1. change the replication factor to exactly match the topology
> > 2. run repair
> > 3. change the replication strategy
> > 
> > there's a hole between 1 and 2, since at that point reads can return incorrect data.
> 
> I think we can avoid the hole by not messing with RF of the old table and repairing the new one instead.
> 
> So the upgrade on each node would look like:
> 
> 1. Copy local mutations to new table
> 2. Perform repair of the new table (so the last node will sync the new table to finally look exactly like the old one)
> 3. Announce feature bit saying "I've copied my data to the new table"
>    (here node still writes to both tables and reads from the old)
> 4. After the cluster feature is on, start reading from the new table and announce feature bit saying "I'm reading from the new table"
> 5. After this feature is on, stop writing to the old table
> 
> Steps 1 and 2 can be done without any cluster feature since in only touches local data and there is no need to wait for other nodes.

It should work, but it's very involved. If we do something this complicated, let's switch it to raft group 0 instead. It should also be easier to coordinate exactly when we switch the reads.

Since it's a small table, we don't need anything complicated like repair. We can read the entire table into memory and use it to seed the new table.

**Comment by Jadw1** [comment]
There is still an ongoing conversation whether should this upgrade be done directly to raft-managed table (ask @eliransin for specifics), so I'm turning this PR into the draft.

**Comment by mykaul** [comment]
> There is still an ongoing conversation whether should this upgrade be done directly to raft-managed table (ask @eliransin for specifics), so I'm turning this PR into the draft.

@Jadw1 - any updates?

**Comment by Jadw1** [comment]
> > There is still an ongoing conversation whether should this upgrade be done directly to raft-managed table (ask @eliransin for specifics), so I'm turning this PR into the draft.
> 
> @Jadw1 - any updates?

AFAIR @eliransin agreed to do the move to the raft but we didn't talked about it recently, so @eliransin knows more than me

**Comment by mykaul** [comment]
> > > There is still an ongoing conversation whether should this upgrade be done directly to raft-managed table (ask @eliransin for specifics), so I'm turning this PR into the draft.
> > 
> > 
> > @Jadw1 - any updates?
> 
> AFAIR @eliransin agreed to do the move to the raft but we didn't talked about it recently, so @eliransin knows more than me

It all depends on the effort:
1. How far off is this patch, vs just doing it over Raft
2. If we complete this feature, then we need both move it to Raft and possibly do the 'upgrade'. What's the cost for this?

**Comment by mykaul** [comment]
@kostja - is this PR still relevant?

**Comment by kostja** [comment]
No, it's not, @Jadw1 please close.

**Pull request closed without merging**